piker/brokers/binance.py

```diff
@@ -0,0 +1,425 @@
+        return resproc(resp, log)
+
+    async def symbol_info(
+        self,
+        sym: Optional[str] = None
+    ):
+        resp = await self._api('exchangeInfo', {})
+        if sym is not None:
+            for sym_info in resp['symbols']:
+                if sym_info['symbol'] == sym:
+                    return sym_info
+            else:
+                raise BrokerError(f'{sym} not found')
+        else:
+            return resp['symbols']
+
+    async def bars(
+        self,
+        symbol: str = 'BTCUSDT',
+        start_time: int = None,
+        end_time: int = None,
+        limit: int = 1000,  # <- max allowed per query
+        as_np: bool = True,
+    ) -> dict:
+        if start_time is None:
+            start_time = binance_timestamp(
+                arrow.utcnow()
+                    .floor('minute')
+                    .shift(minutes=-limit)
+            )
+        
+        if end_time is None:
+            end_time = binance_timestamp(arrow.utcnow())
+
+        bars = await self._api(
+            'klines',
+            {
+                'symbol': symbol,
+                'interval': '1m',
+                'startTime': start_time,
+                'endTime': end_time,
+                'limit': limit
+            }
+        )
+
+        new_bars = [
+            (i,) + tuple(
+                ftype(bar[j])
+                for j, (name, ftype) in enumerate(_ohlc_dtype[1:])
+            ) for i, bar in enumerate(bars)
+        ] 
+
+        array = np.array(new_bars, dtype=_ohlc_dtype) if as_np else bars
+        return array
+
+
+@asynccontextmanager
+async def get_client() -> Client:
+    yield Client()
+
+
+async def stream_messages(ws):
+    while True:
+
+        with trio.move_on_after(5) as cs:
+            msg = await ws.recv_msg()
+
+        # for l1 streams binance doesn't add an event type field so
+        # identify those messages by matching keys
+        if list(msg.keys()) == ['u', 's', 'b', 'B', 'a', 'A']:
+            sym = msg['s']
+            bid = float(msg['b'])
+            bsize = float(msg['B'])
+            ask = float(msg['a'])
+            asize = float(msg['A'])
+
+            yield 'l1', {
+                'symbol': sym,
+                'ticks': [
+                    {'type': 'bid', 'price': bid, 'size': bsize},
+                    {'type': 'bsize', 'price': bid, 'size': bsize},
+                    {'type': 'ask', 'price': ask, 'size': asize},
+                    {'type': 'asize', 'price': ask, 'size': asize}
+                ]
+            }
+
+
+def make_sub(pairs: List[str], sub_name: str, uid: int) -> Dict[str, str]:
+    """Create a request subscription packet dict.
+
+    https://binance-docs.github.io/apidocs/spot/en/#live-subscribing-unsubscribing-to-streams
+    """
+    return {
+        'method': 'SUBSCRIBE',
+        'params': [
+            f'{pair.lower()}@{sub_name}'
+            for pair in pairs
+        ],
+        'id': uid
+    }
+
+
+class AutoReconWs:
```

Comment:
Maybe put this in some sort of `utils` module or something like that, as it is shared between kraken & binance backends.

Reply:
Yah 💯.

We need to fix up the kraken ws respawning anyway so we should probably stick that generic subs mgmt code somewhere as well.